### PR TITLE
Add MapLibreGL

### PR DESCRIPTION
**Pull request by miles-grant-ibigroup**
This PR migrates all overlays to MapLibreGL, and therefore vector tiles.

As the migration is quite significant, the main focus here is to achieve a minimum viable product, while also focusing on making as few changes as possible.

The consequences of this are: 

1. Some features have been removed. Many props have been removed. Most of these props can be re-implemented in the future, either as direct emulations of the previous (sometimes Leaflet-specific) props, or as new ones.
2. Significant refactorings have _not_ been undertaken. For example, many of the marker overlays now simply convert their input data to GeoJSON and then render it. A new architecture should allow to to this more efficiently and to break the code out more easily. However, this would modify the architecture even more than it's currently modified, and so these changes should be done in a separate PR.
3. `transit-vehicle-overlay` has lost all TriMet-specific features. These features were also quite Leaflet-specific, and will therefore also have to be re-implemented from the ground up. The "new" version of the layer is missing some functionality and especially good stories. This may be addressable in this PR.
4. Many styling props still emulate Leaflet. This is another area where an entire separate PR is probably best to decide how best to approach these props.
5. Bounds updating code has been moved into `otp-ui`. Whether this is a good idea, or if perhaps an architecture shift might allow for the de-duplication of this code is another project for a separate PR.

### CODE REVIEW

File: packages/base-map/src/index.tsx [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+  zoom: initZoom = 12
+}: Props): JSX.Element => {
+  const [viewState, setViewState] = React.useState<State>({
+    fitBoundsOptions: {
+      animate: true,
+      duration: 300,
+      essential: false,
+      maxDuration: 600,
+      padding: 200
+    },
```
Comment (binh-dam-ibigroup) [inline packages/base-map/src/index.tsx:66]: Is `fitBoundsOptions` used anywhere? It seems like event `onMove` is the only place where `setViewState` is passed props other than latitude and longitude, and that event does not pass fitting boundaries. `fitBoundsOptions` gets passed around via `onViewportChanged`, but the content seems irrelevant.

File: packages/base-map/src/index.story.tsx [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+                right: 5,
+                top: 10
+              }
+            })
+          // eslint-disable-next-line react/jsx-curly-newline
+        }
+      >
+        Set the bounds
+      </button>
+      <BaseMap passedRef={ref} center={center} forceMaxHeight zoom={17} />
```
Comment (binh-dam-ibigroup) [inline packages/base-map/src/index.story.tsx:91]: I would try not to pass a ref. Instead, move `<MapProvider>` to the storybook level (similar to where we used `<IntlProvider>`) and use hook `useMap` to access the underlying map object. See https://visgl.github.io/react-map-gl/docs/api-reference/use-map.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Good initial porting to maplibre! First round of comments: Among other things, I would stay away from passing refs around, exposing native maplibre props, using "artificial" props( -->`forceMaxHeight`) and removing the `fitBoundsOptions` that doesn't seem to have any use.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Second round of comments. Please use the `useMap` hook correctly, and I think we should not hardcode the map id but expose a prop instead. See other refactor suggestions.

File: packages/vehicle-rental-overlay/package.json [hunk 3]
```diff
@@ -11,27 +11,29 @@
...
+    "tsc": "tsc"
   },
   "repository": {
     "type": "git",
     "url": "git+https://github.com/opentripplanner/otp-ui.git"
   },
   "dependencies": {
+    "@opentripplanner/base-map": "^3.0.0-alpha.1",
     "@opentripplanner/core-utils": "^4.5.0",
+    "react-map-gl": "^7.0.15",
```
Comment (binh-dam-ibigroup) [inline packages/vehicle-rental-overlay/package.json:24]: This is a peer dependency

File: packages/vehicle-rental-overlay/src/StationPopup.tsx [hunk 4]
```diff
@@ -0,0 +1,146 @@
+import {
+  Company,
+  ConfiguredCompany,
+  MapLocationActionArg,
+  Station
+} from "@opentripplanner/types";
+import { Styled as BaseMapStyled } from "@opentripplanner/base-map";
+import React from "react";
+import FromToLocationPicker from "@opentripplanner/from-to-location-picker";
```
Comment (binh-dam-ibigroup) [inline packages/vehicle-rental-overlay/src/StationPopup.tsx:9]: Sort imports by package name

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Last round of comments for now. similar requests as my previous round.

File: packages/base-map/src/styled.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  stroke?: number;
+  strokeColor?: string;
+};
+/**
+ * @deprecated this marker was created to make the transition from Leaflet more manageable,
+ * but in most cases this marker should not be used -- use a MapLibreGL Circle instead
+ *
+ * https://maplibre.org/maplibre-gl-js-docs/style-spec/layers/#circle
+ */
+export const LeafletStyleMarker = styled.div<LeafletStyleMarkerProps>`
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/base-map/src/styled.ts:31]: Yes this needed to draw circles on the map without using GeoJSON. It's marked as deprecated, and can be removed once we refactor away all non-GeoJSON rendering (which is better for performance anyway)

File: packages/base-map/__mocks__/ContextMenuDemo.tsx [hunk 6]
```diff
@@ -0,0 +1,49 @@
+import { LngLat } from "maplibre-gl";
+import React, { Component } from "react";
+import { Popup } from "react-map-gl";
+import BaseMap from "../lib";
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/base-map/__mocks__/ContextMenuDemo.tsx:4]: Good catch thank you!

File: packages/base-map/__mocks__/ContextMenuDemo.tsx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+const center: [number, number] = [45.522862, -122.667837];
+
+type State = {
+  contents?: React.ReactNode;
+  location?: [number, number];
+};
+type Props = Record<string, never>;
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/base-map/__mocks__/ContextMenuDemo.tsx:12]: As far as I can tell, `Record<string, never>` is the best way to do an empty type in TS

File: packages/base-map/src/index.story.tsx [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+      <MarkerWithPopup position={[center[0], center[1]]} />
+      <MarkerWithPopup position={[center[0] + 0.01, center[1]]} />
+    </LayerWrapper>
+    <AllVehiclesOverlay id="layer-2" />
+  </BaseMap>
+);
+// Custom styling for this story only, not in production
+customLocationPopupContent.parameters = a11yOverrideParameters;
+
+export const onContextMenuPopup = () => <ContextMenuDemo />;
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/base-map/src/index.story.tsx:148]: It being in its own file is a holdover from Leaflet, but even now the `onContextMenu` handler on the `BaseMap` means that this would involve moving the entire example to this file which I'm not crazy about.

File: packages/base-map/src/index.tsx [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+ *
+ * Overlays are automatically added to the overlay control displayed by the
+ * BaseMap. The user uses that control to turn overlays on or off. Only overlays
+ * with an id are added to the control.
+ */
+type Props = React.ComponentPropsWithoutRef<React.ElementType> & {
+  baseLayer?: string;
+  center?: [number, number];
+  forceMaxHeight?: boolean;
+  mapLibreProps?: MapProps;
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/base-map/src/index.tsx:28]: They're also used by otp-rr: https://github.com/opentripplanner/otp-react-redux/compare/dev...vector-tiles#diff-3a3b7fe3787ef4562afac83172fbe1e14091715b7897e6ea60ff4346a6b0bb86R280

File: packages/base-map/src/index.story.tsx [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+ */
+export const withOptionalControls = () => (
+  <BaseMap
+    center={center}
+    forceMaxHeight
+    // We supply our own AttributionControl, so disable the default one
+    // See https://visgl.github.io/react-map-gl/docs/api-reference/attribution-control
+    mapLibreProps={{ attributionControl: false }}
+  >
+    <AttributionControl customAttribution="This adds to the attribution information supplied by the map style json" />
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/base-map/src/index.story.tsx:165]: In addition to being used by otp-rr, I'm anticipating that exposed `mapLibreProps` will probably be welcomed by others implementing otp-ui

File: packages/base-map/src/index.tsx [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+        onContextMenu={onContextMenu}
+        onMove={evt => setViewState(evt.viewState)}
+        onTouchStart={() => {
+          setFakeMobileHover(false);
+        }}
+        style={{
+          display: "block",
+          height: forceMaxHeight ? "90vh" : "100%",
+          width: "100%"
+        }}
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/base-map/src/index.tsx:136]: I wanted to avoid repeating the same inline style in the stories, but story-based styled component might work... 

`100%` doesn't work in the storybook unfortunately.

File: packages/base-map/src/index.tsx [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+  useEffect(() => {
+    if (center?.[0] === null || center?.[1] === null) return;
+
+    setViewState({
+      ...viewState,
+      latitude: center?.[0],
+      longitude: center?.[1]
+    });
+  }, [center]);
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/base-map/src/index.tsx:85]: I might be mistaken, but wouldn't this cause an infinite loop where `setViewState` is called every time `viewState` is updated?

File: packages/base-map/src/index.tsx [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+            onTouchEnd={() => {
+              setFakeMobileHover(true);
+            }}
+            className="filter-group"
+            id="filter-group"
+          >
+            <ul
+              className={`layers-list ${fakeMobileHover &&
+                "fake-mobile-hover"}`}
+            >
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/base-map/src/index.tsx:150]: This isn't possible because the `layers-list` pseudo elements need to be rendered inside the box

File: packages/park-and-ride-overlay/src/ParkAndRideOverlay.story.js [hunk 14]
```diff
@@ -5,7 +5,7 @@ import { action } from "@storybook/addon-actions";
 import ParkAndRideOverlay from ".";
 import parkAndRideLocations from "../__mocks__/park-and-rides.json";
 
-import "../../../node_modules/leaflet/dist/leaflet.css";
+import "maplibre-gl/dist/maplibre-gl.css";
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/park-and-ride-overlay/src/ParkAndRideOverlay.story.js:8]: I kept it since it seemed to be required with Leaflet, but it doesn't seem to be required anymore. I will remove it

File: packages/base-map/src/index.tsx (hunk 1 above)
Comment (miles-grant-ibigroup) [reply packages/base-map/src/index.tsx:66]: These options would normally be passed in via props to MapLibreGL, but since we're using a controlled component we must set these in state. It's not possible to have some elements of viewState be controlled and some uncontrolled.

File: packages/endpoints-overlay/src/endpoint.tsx [hunk 15]
```diff
@@ -92,22 +79,27 @@ function UserLocationIcon({ type }: IconProps) {
...
   };
 }
 
-class Endpoint extends Component<Props> {
-  rememberAsHome = (): void => {
-    const { location: propsLocation, rememberPlace } = this.props;
+const Endpoint = (props: Props): JSX.Element => {
+  const intl = useIntl();
+
+  const rememberAsHome = (): void => {
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/endpoints-overlay/src/endpoint.tsx:101]: Do you mind if I move this to a new issue? This doesn't feel related to this PR, which is already quite large.

File: packages/route-viewer-overlay/src/index.tsx [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+ */
+const removePointsInFlexZone = (stops: Stop[], points: [number, number][]) => {
+  // First, go through all stops to find flex zones
+  const bboxes =
+    stops
+      ?.map(stop => {
+        if (stop.geometries?.geoJson?.type !== "Polygon") {
+          return null;
+        }
+        return stop.geometries.geoJson.coordinates?.[0] || null;
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/route-viewer-overlay/src/index.tsx:37]: Wouldn't this result in more passes over the array? Normally I'd make the sacrifice for readability but this method gets called very often and blocks render

[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/route-viewer-overlay/src/index.tsx:37]: If anything, should we remove that second filter and merge it into the map?

File: packages/route-viewer-overlay/src/index.tsx [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+        if (props.mapCenterCallback) {
+          props.mapCenterCallback();
+        }
+      }
+    }
+  }, [routeData.patterns]);
+
+  const { clipToPatternStops, path } = props;
+
+  if (!routeData || !routeData.patterns) return <></>;
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/route-viewer-overlay/src/index.tsx:87]: MapLibreGL seems to require a non-null value being returned

File: packages/route-viewer-overlay/src/index.tsx [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+  });
+
+  const geojson: GeoJSON.FeatureCollection = {
+    type: "FeatureCollection",
+    features: segments.map(segment => ({
+      type: "Feature",
+      properties: [],
+      geometry: {
+        type: "LineString",
+        coordinates: segment
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/route-viewer-overlay/src/index.tsx:110]: For GeoJSON isn't it more readable to have the type first? Most GeoJSON I've seen in the wild contains the type first

[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/route-viewer-overlay/src/index.tsx:110]: I've swapped properties and geometry

File: packages/base-map/src/index.tsx [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+    [hiddenLayers]
+  );
+
+  return (
+    <MapProvider>
+      <Map
+        // eslint-disable-next-line react/jsx-props-no-spreading
+        {...mapLibreProps}
+        ref={passedRef}
+        id="mainMap"
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/base-map/src/index.tsx]: I've removed the id prop

File: packages/stops-overlay/src/index.tsx [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+    STOP_LAYERS.forEach(stopLayer => {
+      mainMap?.on("mouseenter", stopLayer, () => {
+        mainMap.getCanvas().style.cursor = "pointer";
+      });
+      mainMap?.on("mouseleave", stopLayer, () => {
+        mainMap.getCanvas().style.cursor = "";
+      });
+      mainMap?.on("click", stopLayer, (event: EventData) => {
+        setClickedStop(event.features?.[0].properties);
+      });
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/stops-overlay/src/index.tsx]: In a future refactor, I'd like to see one single click handler!

File: packages/zoom-based-markers/src/ZoomBasedMarker.story.tsx [hunk 21]
```diff
@@ -1,19 +1,23 @@
 // Removed as core-utils is typescripted. TODO: Remove when typescripting!
 /* eslint-disable react/forbid-prop-types */
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/zoom-based-markers/src/ZoomBasedMarker.story.tsx:2]: I've adjusted the reasoning for it

File: packages/stop-viewer-overlay/src/index.tsx [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+  /**
+   * Only reset map view if a new stop is selected. This prevents resetting the
+   * bounds if, for example, the arrival times have changed for the same stop
+   * in the viewer.
+   */
+  useEffect(() => {
+    mainMap?.flyTo({ center: [stop.lon, stop.lat] });
+  }, [stop.lat, stop.lon]);
+
+  if (visible === false || !stop || !StopMarker) return <></>;
```
[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply packages/stop-viewer-overlay/src/index.tsx:30]: Can you add what you said in a comment?

File: packages/endpoints-overlay/src/endpoint.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply packages/endpoints-overlay/src/endpoint.tsx:101]: Or you can revert to a class

**Comment by miles-grant-ibigroup** [comment]
Thanks Binh for all your comments! Some very good catches in there. I've addressed all resolved comments, otherwise I've left a note.

File: packages/route-viewer-overlay/src/index.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply packages/route-viewer-overlay/src/index.tsx:36]: I'd separate the map and the filtering, unless there is some benchmark that justifies otherwise (in that case, add a comment with the results of such benchmark)

File: packages/base-map/src/index.tsx [hunk 23]
```diff
@@ -22,24 +22,34 @@ import MarkerWithPopup from "./MarkerWithPopup";
  * with an id are added to the control.
  */
 type Props = React.ComponentPropsWithoutRef<React.ElementType> & {
+  /** A URL pointing to the vector tile specification which should be used as the main map.  */
   baseLayer?: string;
+  /** A [lat, lon] position to center the map at. */
   center?: [number, number];
+  /** Whether or not to override the default MapLibre 100% height to be 90vh */
   forceMaxHeight?: boolean;
```
Comment (binh-dam-ibigroup) [inline packages/base-map/src/index.tsx:30]: I REALLY don't like this prop and it should definitely be removed, but I know what is happening.

In storybook, previously, a style tag was inserted that set the leaflet container height to 480px. 
I think we still want to set some height, so that the map shows in storybook. As written, the map is set to take 100% of the height, but no height is defined in the parent tags so basically the height is zero. This is not the case and won't be an issue in OTP-RR where the height is defined by the successive map containers.

What we should do here is to use the `className` and `style` props. In storybook, define a `styled(BaseMap)` with the `height` of 480px, 90vh or whatever and render the styled component in each story.

File: packages/base-map/src/index.tsx [hunk 24]
```diff
@@ -112,68 +125,64 @@ const BaseMap = ({
...
+      onMove={evt => setViewState(evt.viewState)}
+      onTouchStart={() => {
+        setFakeMobileHover(false);
+      }}
+      ref={passedRef}
+      style={{
+        display: "block",
+        height: forceMaxHeight ? "90vh" : "100%",
+        width: "100%"
+      }}
```
Comment (binh-dam-ibigroup) [inline packages/base-map/src/index.tsx:147]: Per my comment about the `forceMaxHeight` prop, pass in the `style` prop instead. Also pass the `className` prop so that styles can be overwritten by CSS or by the style prop. This component renders as a `<div>`, so by default it's `display: block`, which sets the `width` to 100% already.

File: packages/base-map/src/index.story.tsx (hunk 2 above)
Comment (binh-dam-ibigroup) [reply packages/base-map/src/index.story.tsx:91]: Surround the `<div>` or replace it with a `<MapProvider>`, give an `id` to the map (see also my comment about passing in an `id` prop), and you will be able to use `useMap`, so we can get rid of the `passedRef` prop.

File: packages/base-map/src/index.tsx [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+      setHiddenLayers(updatedLayers);
+    },
+    [hiddenLayers]
+  );
+
+  return (
+    <Map
+      // eslint-disable-next-line react/jsx-props-no-spreading
+      {...mapLibreProps}
+      latitude={viewState.latitude}
```
Comment (binh-dam-ibigroup) [inline packages/base-map/src/index.tsx:131]: Add and pass in an `id` prop, so that implementers can use the `useMap` hook instead of passing refs around.

File: packages/base-map/src/index.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply packages/base-map/src/index.tsx:32]: Should `reuseMap` just be set to true, or exposed as a prop?

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
See suggestions regarding the `passedRef` and `forceMaxHeight` props, which should make the component cleaner and easier to customize. I'm still on the fence with the `maplibreProps` props.

File: packages/trip-viewer-overlay/package.json [hunk 26]
```diff
@@ -21,7 +21,7 @@
   "dependencies": {
     "@mapbox/polyline": "^1.1.0",
     "@opentripplanner/base-map": "^3.0.0-alpha.1",
-    "@opentripplanner/core-utils": "^4.5.0"
+    "@opentripplanner/core-utils": "^6.0.0"
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/trip-viewer-overlay/package.json:24]: Don't see why not! This is a good opportunity

File: packages/base-map/src/index.tsx (hunk 25 above)
Comment (miles-grant-ibigroup) [reply packages/base-map/src/index.tsx:131]: Even with an id, `useMap` only works when the component is a child of the parent. The id is only useful when managing multiple maps.

**Comment by miles-grant-ibigroup** [comment]
I don't like exposing `reuseMap` set to true by default because it has some strange side-effects. I also don't like exposing MapLibre props on our own, as that makes the entire `BaseMap` a bit redundant, no?

**Comment by miles-grant-ibigroup** [comment]
Thanks for the review -- I've addressed all changes other than the `forceMaxHeight` issue. I had already tried the approaches you suggested before landing on the `forceMaxHeight` solution. There is definitely a better solution out there, but we will need to discuss it offline

File: packages/base-map/src/index.tsx (hunk 25 above)
Comment (binh-dam-ibigroup) [reply packages/base-map/src/index.tsx:131]: See these docs for `useMap` outside of a `<Map>` element: https://visgl.github.io/react-map-gl/docs/api-reference/use-map

Comment (miles-grant-ibigroup) [reply packages/base-map/src/index.tsx:131]: In that example the `NavigateButton` is a child of a `MapProvider` that the Map is also a child of. This architecture doesn't work in our case (I experimented with this which is why that `MapProvider` was originally left behind)

File: packages/transitive-overlay/src/TransitiveOverlay.story.js [hunk 27]
```diff
@@ -63,13 +63,22 @@ function getCustomRouteLabel(itineraryLeg) {
   return null; // null or undefined or empty string will tell transitive-js not to render a route label
 }
 
+const withMapContainer = Story => (
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/transitive-overlay/src/TransitiveOverlay.story.js:66]: Do you know if it's possible to pass props to the decorator? The reason the `BaseMap` is repeated here is because the `center` changes

File: packages/endpoints-overlay/src/EndpointsOverlay.story.tsx [hunk 28]
```diff
@@ -44,11 +44,13 @@ function CatDogIcon({ type }: UserLocationAndType) {
...
+  <BaseMapStyled.StoryMapContainer>
+    <BaseMap center={center} zoom={zoom}>
+      {/* For some reason, <Story /> does not work with snapshots,
         so use the function syntax instead. */}
-    {Story()}
-  </BaseMap>
+      {Story()}
+    </BaseMap>
+  </BaseMapStyled.StoryMapContainer>
 );
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/endpoints-overlay/src/EndpointsOverlay.story.tsx]: I've implemented this with [e70ca8b](https://github.com/opentripplanner/otp-ui/pull/420/commits/e70ca8b4344feea1d841a1f24173c80248c582fd)

**Comment by miles-grant-ibigroup** [comment]
Thanks for the suggestion Binh, it seems to work great!

File: packages/trip-viewer-overlay/src/TripViewerOverlay.story.tsx [hunk 29]
```diff
@@ -1,37 +1,30 @@
-import BaseMap, { Styled as BaseMapStyled } from "@opentripplanner/base-map";
 import React from "react";
 
 import tripData from "../__mocks__/mock-trip.json";
 import TripViewerOverlay from ".";
+import { withMap } from "../../../.storybook/base-map-wrapper";
```
Comment (binh-dam-ibigroup) [inline packages/trip-viewer-overlay/src/TripViewerOverlay.story.tsx:5]: I have not checked the other stories, but the `import ... from "."` should be below this one.

File: packages/vehicle-rental-overlay/src/VehicleRentalOverlay.story.tsx [hunk 30]
```diff
@@ -51,22 +51,16 @@ const ZoomControlledMapWithVehicleRentalOverlay = ({
...
-      />
-    </BaseMap>
-  </BaseMapStyled.StoryMapContainer>
+  <VehicleRentalOverlay
+    id="test"
+    configCompanies={configCompanies}
+    companies={companies}
+    getStationName={getStationName}
+    setLocation={setLocation}
+    refreshVehicles={refreshVehicles}
```
Comment (binh-dam-ibigroup) [inline packages/vehicle-rental-overlay/src/VehicleRentalOverlay.story.tsx:60]: Sort props

File: packages/vehicle-rental-overlay/src/index.tsx [hunk 31]
```diff
@@ -6,8 +6,8 @@ import {
   Station
 } from "@opentripplanner/types";
 import { EventData } from "mapbox-gl";
-import { useMap, Layer, Popup, Source } from "react-map-gl";
 import React, { useEffect, useState } from "react";
+import { useMap, Layer, Popup, Source } from "react-map-gl";
```
Comment (binh-dam-ibigroup) [inline packages/vehicle-rental-overlay/src/index.tsx:10]: Sort imports

File: packages/base-map/src/index.story.tsx [hunk 32]
```diff
@@ -129,7 +120,9 @@ export const withSampleMarkers = () => (
 
 export const overlayWithLargeDataSet = () => (
   <>
-    <div>Do not add Storybook overhead on layers with large dataset...</div>
+    <div style={{ position: "relative", zIndex: 1000 }}>
```
Comment (binh-dam-ibigroup) [inline packages/base-map/src/index.story.tsx:123]: The "Click-to-set-bounds" story needs something like this too, and probably you need to say for that story and this one that they need no decorator (otherwise, two maps get rendered!)

File: packages/base-map/src/index.story.tsx [hunk 33]
```diff
@@ -0,0 +1,175 @@
...
+export const customLocationPopupContent = () => (
+  <BaseMap center={center}>
+    <Popup longitude={center[1]} latitude={center[0]}>
+      {samplePopup}
+    </Popup>
+  </BaseMap>
+);
+export const optionalLayers = () => (
+  <BaseMap center={center}>
+    <LayerWrapper
```
Comment (binh-dam-ibigroup) [inline packages/base-map/src/index.story.tsx:141]: Just to check, is it intended that you can activate the hidden layer but not hide it again?

**Review by binh-dam-ibigroup — APPROVED** [review]
It's much cleaner after the latest changes! There are some double map issues to fix from creating the map decorator, and some minor formatting things, but otherwise nice work!

File: packages/base-map/src/index.story.tsx (hunk 33 above)
Comment (miles-grant-ibigroup) [reply packages/base-map/src/index.story.tsx:141]: No! This is a regression from when we moved the hiddenLayer updating to `useCallback`. I've reverted this change. The problem was the state that the `useCallback` method depended on was changing from inside the callback, so the callback never fired the update.

Comment (miles-grant-ibigroup) [reply packages/base-map/src/index.story.tsx]: Thanks for catching that

**Comment by miles-grant-ibigroup** [comment]
This is ready for your review @demory

File: packages/base-map/package.json [hunk 34]
```diff
@@ -9,20 +9,22 @@
   "module": "esm/index.js",
   "private": false,
   "dependencies": {
-    "prop-types": "^15.7.2"
+    "@opentripplanner/types": "^2.0.0",
```
Comment (binh-dam-ibigroup) [inline packages/base-map/package.json:12]: Could you use types 3.0.0 please? Oh, it is a dev dependency.

Comment (miles-grant-ibigroup) [reply packages/base-map/package.json]: Good call thanks have updated

**Comment by miles-grant-ibigroup** [comment]
Merge conflicts are resolved now, types and core-utils are updated across all packages.

**Comment by demory** [comment]
@miles-grant-ibigroup something I've noticed a couple times in my testing -- in any overlay with a collection of items (stops, vehicles e.g.) *and* popups for the items, clicking on one item and then on another hides the first popup, but doesn't show the 2nd popup. You have to click the 2nd thing again to show its popup.

This is different from the current behavior in leaflet / otp-rr, where one click both hides the 1st popup and shows the 2nd. Is that something that can be recreated here via configuration, or would it need code changes?

**Comment by demory** [comment]
Another popup comment: in the StopsOverlay stories, with the "Default" story, opening a popup and then zooming keeps the popup displayed. But with "Flex Stops", the popup disappears after a zoom change. What accounts for that?

**Comment by miles-grant-ibigroup** [comment]
> This is different from the current behavior in leaflet / otp-rr, where one click both hides the 1st popup and shows the 2nd. Is that something that can be recreated here via configuration, or would it need code changes?

This seems to be related to the layer event handling. It would be possible to replicate the old behavior, but we'd have to overwrite MapLibre's map event handler. It's something that's possible, but if we wanted to go forward with it I'd do so in a new PR.

> Another popup comment: in the StopsOverlay stories, with the "Default" story, opening a popup and then zooming keeps the popup displayed. But with "Flex Stops", the popup disappears after a zoom change. What accounts for that?

This has to do with the fact that there are multiple layers being rendered at the same time. Each layer has its own event handler and in the flex stops there are a lot that are competing to figure out what's going on. I would see what the behavior is like inside otp-rr before debugging anything too strongly. Again this can be solved with a custom event handler, but I would save that for a separate PR once we see how this behaves inside otp-rr.

> This is fairly minor and maybe we come back to it in the future, but with the line labeling is there an (easy-ish) way to recreate the current functionality, with the label always horizontally aligned inside a roundel or similar? At least for short (1-3 character) labels

There are a _lot_ of text options inside MapLibre. So far I've enabled one that ensures that the text is never upside down. There seem to be others that might do what we want, I'll fiddle around with it some more. That being said, I think it's probably wiser to not spend too much time trying to replicate transitive behavior here and instead work towards getting transitive to render to MapLibre. Once we add webGL as a target renderer for transitive, it should be quite trivial to display it on the map: https://maplibre.org/maplibre-gl-js-docs/example/custom-style-layer/

File: packages/transitive-overlay/package.json [hunk 35]
```diff
@@ -11,24 +11,28 @@
...
     "url": "git+https://github.com/opentripplanner/otp-ui.git"
   },
   "dependencies": {
+    "@mapbox/polyline": "^1.1.1",
+    "@opentripplanner/base-map": "^3.0.0-alpha.3",
     "@opentripplanner/itinerary-body": "^4.0.0",
     "@opentripplanner/core-utils": "^7.0.0",
     "lodash.isequal": "^4.5.0",
-    "transitive-js": "^0.14.1"
+    "transitive-js": "^0.14.1",
```
Comment (demory) [inline packages/transitive-overlay/package.json:27]: Is this still actually used?

Comment (miles-grant-ibigroup) [reply packages/transitive-overlay/package.json:27]: Good catch! I was being optimistic for the future but yes at the moment this is unused. I've removed it in 126e4300e171b9c3f7b1da89a3c885e544279db4

Comment (demory) [reply packages/transitive-overlay/package.json]: Thanks! That will definitely cut down the bundle size and by the time we re-add transitive hopefully it will be much smaller, i.e. w/o the d3 dependency, etc.

**Review by demory — APPROVED** [review]
